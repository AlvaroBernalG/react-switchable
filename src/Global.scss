/*  Switch -> The container

    *------------------------*
    |                        |
    |                        |
    *------------------------*
*/

$switch_width: 610px;
$switch_height: 100%;
$switch_min_height: 45px;
$switch_margin: 10px;
$switch_border_radius: 100px;
$switch_border_color: white;

/*  State -> Every Item inside the container 

    *------*-----------------*
    | Item |                 |
    |      |                 |
    *------*-----------------*
*/

$state_width: 100%;
$state_height: 100%;
$state_padding: 15px;
$state_color: inherit;
$state_fontFamily: 'Muli';
$state_letter_spacing: 1.2px;
$state_font_weight: bold;
$state_color_hover: darken(white, 10%);
$state_background: transparent;

// On Active:
$state_color_onActive: #e66549;

/*   Overlay -> The layer that covers every item when item is active.

    *------*-----------------*
    | On   |                 |
    |      |                 |
    *------*-----------------*
*/

$overlay_height: 100%;
$overlay_background_color: rgba(249, 249, 249, 0.5);
$overlay_border_radius: $switch_border_radius;
$overlay_box_shadow: 0 0 6px 1px #e66549;
$overlay_transition: all 0.4s ease;

// On Active:

// State modifiers
