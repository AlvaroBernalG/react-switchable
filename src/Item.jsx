import React from "react";
import PropTypes from "prop-types";
import "./Item.scss";

const Item = props => {
  Item.propTypes = {
    value: PropTypes.string.isRequired,
    children: PropTypes.node,
    disable: PropTypes.bool,
    active: PropTypes.bool,
    tabIndex: PropTypes.number,
    className: PropTypes.string
  };

  Item.defaultProps = {
    children: "",
    active: false,
    disable: false,
    tabIndex: 0,
    className: ""
  };

  const { active, className, disable, value, ...rest } = props;
  // console.log("Should be disable", props, disable);

  const classes = [
    "abg-switch__item",
    className,
    active ? "abg-switch__item--on" : "",
    disable ? "abg-switch__item--disable" : ""
  ].join(" ");

  return (
    <span
      className={classes}
      {...rest}
      role="radio"
      aria-checked={active}
      aria-disabled={disable}
    >
      {props.children}
      <input type="radio" value={value} defaultChecked={active} />
    </span>
  );
};

export default Item;
