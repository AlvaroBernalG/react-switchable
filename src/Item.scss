@import 'Global';
.abg-switch__item {
  display: flex;
  justify-content: center;
  align-items: center;
  letter-spacing: $item_letter_spacing;
  cursor: pointer;
  width: $item_width;
  height: $item_height;
  padding: $item_padding;
  font-weight: $item_font_weight;
  background: $item_background;
  overflow: visible;
  color: inherit;
  font: inherit;
  border: none;
  outline: none;
  &:not(.switch__item--disable):hover {
    color: $item_color_hover;
  }
  input {
    display: none;
  }
}

.abg-switch__item.abg-switch__item--on {
  z-index: 2;
  color: $item_color_onActive;
  font-weight: bold;
  outline: none;
 &:hover {
    color: $item_color_onActive;
  }
}

.abg-switch__item.abg-switch__item--disable {
  opacity: $item_disable_opacity;
  &:hover {
    color: inherit;
  }
}
