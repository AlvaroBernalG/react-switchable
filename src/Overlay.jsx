import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import './Overlay.scss'

const getOverlayClasses = classTransition =>
  classNames('switch__overlay', ([classTransition]: true))

const Overlay = props => {
  const { goTo, items, classTransition, ...rest } = props
  return (
    <span
      {...rest}
      draggable
      onDragStart={() => this.onDragStart()}
      onDragEnd={() => this.onDragEnd()}
      className={getOverlayClasses(classTransition)}
      style={{
        width: `${100 / items}%`,
        transform: `translateX(${100 * goTo}%)`
      }}
    />
  )
}

Overlay.propTypes = {
  goTo: PropTypes.number.isRequired,
  items: PropTypes.number.isRequired,
  classTransition: PropTypes.string
}

Overlay.defaultProps = {
  classTransition: 'switch__overlay-transition'
}

export default Overlay
