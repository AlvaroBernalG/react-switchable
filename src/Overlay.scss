@import 'Global';
.abg-switch__overlay {
  position: absolute;
  height: $overlay_height;
  border-radius: $overlay_border_radius;
  background-color: $overlay_background_color;
  box-shadow: $overlay_box_shadow;
  border: 1px solid transparent;
  box-sizing: border-box;
  transition: $overlay_transition;
  outline: none;
}

