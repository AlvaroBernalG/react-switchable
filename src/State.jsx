import React from "react";
import classNames from "classnames";
import PropTypes from "prop-types";
import "./State.scss";

const State = props => {
  const { active, tabIndex, className, disable, value, ...rest } = props;
  return (
    <span
      {...rest}
      role="radio"
      aria-checked={active}
      tabIndex={tabIndex}
      className={classNames("abg-switch__state", [className], {
        "abg-switch__state--on": active,
        "abg-switch__state--disable": disable
      })}
    >
      {props.children}
      <input type="radio" value={value} checked={active} />
    </span>
  );
};

State.propTypes = {
  value: PropTypes.string.isRequired,
  children: PropTypes.node,
  disable: PropTypes.bool,
  active: PropTypes.bool,
  tabIndex: PropTypes.number,
  className: PropTypes.string
};

State.defaultProps = {
  children: "",
  active: false,
  disable: false,
  tabIndex: 0,
  className: ""
};

export default State;
