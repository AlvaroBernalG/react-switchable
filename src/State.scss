@import 'Global';
.abg-switch__state {
  display: flex;
  justify-content: center;
  align-items: center;
  letter-spacing: $state_letter_spacing;
  cursor: pointer;
  width: $state_width;
  height: $state_height;
  padding: $state_padding;
  font-weight: $state_font_weight;
  background: $state_background;
  overflow: visible;
  color: inherit;
  font: inherit;
  border: none;
  &:not(.switch__state--disable):hover {
    color: $state_color_hover;
  }
}

.abg-switch__state--on {
  z-index: 2;
  color: $state_color_onActive !important;
  font-weight: bold;
}

.abg-switch__state--disable {}
