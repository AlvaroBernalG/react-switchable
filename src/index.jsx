export { default } from "./Switch";
export { default as State } from "./State";
export { default as Overlay } from "./Overlay";
export { default as Item } from "./State";
