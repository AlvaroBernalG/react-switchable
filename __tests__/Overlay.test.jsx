import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import { Overlay } from "../src/";

describe("<Overlay />", () => {
  it("matches the snapshot", () => {
    const tree = shallow(<Overlay goTo={0} items={2} />);
    expect(tree).toMatchSnapshot();
  });
});
