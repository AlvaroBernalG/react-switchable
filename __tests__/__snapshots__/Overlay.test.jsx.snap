// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay /> matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Overlay
    className=""
    classTransition="abg-switch__overlay-transition"
    goTo={0}
    items={2}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "className": " abg-switch__overlay",
      "style": Object {
        "transform": "translateX(0%)",
        "width": "50%",
      },
    },
    "ref": null,
    "rendered": null,
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "className": " abg-switch__overlay",
        "style": Object {
          "transform": "translateX(0%)",
          "width": "50%",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
